Issue 4426: Transaction performance metrics

RaulGracia opened the pull request:
**Change log description**  
Added and refactored transaction-related metrics to understand better transaction performance.

**Purpose of the change**  
Fixes #4426.

**What the code does**  
Currently, the performance of Transactions is measured only at the `ControllerService` level. This is an insufficient approach to measure the actual performance of Transactions. This PR is aimed at improving the current metrics we export of Transactions in the following ways:
- Distinguish between `COMMITTING` and `COMMITTED` Transactions: Currently, the only notion of latency we have of a Transaction is the time it takes to get from `OPEN` to `COMMITTING` or `ABORTING` (`CREATED` is fine as it is as sync operation). These metrics can be easily understood as the time it takes for a Transaction to get committed/aborted, but this is not true. For this reason, this PR renames the metrics reported at the `ControllerService` level as `[COMMITTING|ABORTING]_TRANSACTION_LATENCY`. Now, the `ControllerService` reports the latency it takes for switching Transactions to committing/aborting states.

- Measure the actual time to commit/abort Transactions in their respective handlers: The actual processes that perform the background job of committing and aborting transactions are their respective request handlers. For this reason, the `COMMIT` and `ABORT` Transaction latency now are measured at this level.

- Separate metric for Segment management: Moreover, we have added a special kind of metrics related to the time it takes to the Controller to manage the segments against the Segment Store for a particular transaction (merge, delete, create,...). In case of low Transaction performance, this metric can tell us whether the problem is with the Segment manegament or with the metadata management of Transactions.


**How to verify it**  
We have added an integration test to check the publishing of Transactions metrics. All other tests should be passing as before. Also, we have deployed a Pravega cluster with these changes and the metrics are being reported as expected:

![new-transaction-dashboard](https://user-images.githubusercontent.com/717112/70633107-492e4b00-1c30-11ea-82f0-6610de7354cb.png)

shiveshr commented on controller/src/main/java/io/pravega/controller/metrics/TransactionMetrics.java:
```diff
@@ -10,19 +10,10 @@
-import static io.pravega.shared.MetricsNames.CREATE_TRANSACTION_LATENCY;
-import static io.pravega.shared.MetricsNames.OPEN_TRANSACTIONS;
-import static io.pravega.shared.MetricsNames.globalMetricName;
+import static io.pravega.shared.MetricsNames.*;
```
nit: please revert it

RaulGracia replied:
Done.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessors.java:
```diff
@@ -100,23 +101,25 @@ public ControllerEventProcessors(final String host,
                                      final StreamTransactionMetadataTasks streamTransactionMetadataTasks,
-                                     final ScheduledExecutorService executor) {
+                                     final ScheduledExecutorService executor,
+                                     final TransactionMetrics transactionMetrics) {
```
Should we pass new transactionMetrics object here? Can we not do it like other metrics, where we use static provider and get specific transaction metrics?

RaulGracia replied:
I see two options:
1) We pass the metrics object as currently done, so we need to modify the constructor of all classes using it.
2) We make `TransactionMetrics` (and maybe `StreamMetrics` too) singleton objects, so we only need to invoke something like `TransactionMetrics.getInstance().xxxx()` from the places where we want to report metrics. This would not require to change the constructors for that class.

Please, let me know your opinion on that.

shiveshr replied:
My preference would be to use state/singleton objects. So let us focus on `2` and see if it is a viable solution. 
Having static/singleton objects exposed via static getters, they can be accessed from anywhere without having to pass these objects in constructors. 
However, do remember that controller modules can be reinitialized (in case of zk session expiration, all modules restart internally).
We need to assess if singleton metrics objects could have any detrimental effect in any way. Let me elaborate - if there was any state (any metrics) in the previous iteration of metrics object, do we need to reset such states during restarts of other modules that use this metrics object?

If not, we can use singleton objects without any worry.  
However, if we need to instantiate new metrics object each time we restart, then we could have these metrics initialized in ControllerServiceStarter and reinitialized upon each restart which refreshes their state or creates a new object altogether perhaps.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java:
```diff
@@ -106,12 +113,13 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
                             log.info("Successfully committed transactions on epoch {} on stream {}/{}", r, scope, stream);
                         } else {
-                            log.info("No transactions found in committing state on stream {}/{}", r, scope, stream);
+                            log.info("No transactions found in committing state {} on stream {}/{}", r, scope, stream);
```
please replace this with
`                             log.info("No transactions found in committing state on stream {}/{}", scope, stream); `

RaulGracia replied:
Fixed.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java:
```diff
@@ -180,10 +189,10 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
                                                     // If active epoch's reference is same as transaction epoch,
                                                     // we can commit transactions immediately
-                                                    return commitTransactions(scope, stream, new ArrayList<>(activeEpochRecord.getSegmentIds()), txnList, context)
+                                                    return commitTransactions(scope, stream, new ArrayList<>(activeEpochRecord.getSegmentIds()), txnList, context, timer)
```
we dont need to pass timer to all methods, esp when its only exercised at one place after the last step.
 we should instead timer.getElapsed at the end of this method itself.

RaulGracia replied:
I see two problem with that approach:
1) Granularity of transaction latency report: If we just get the time elapsed after this method, we are not measuring the time to commit individual transactions, but the whole batch. In the PR, we execute a `timer.getElapsed()` for each transaction, which looks more precise to me.
2) Completeness: I realized that the `rollTransactions()` method also invokes internally `commitTransactions()` (within `copyTxnEpochSegmentsAndCommitTxns()`). This means that we also need to cover this path to report all the transaction commit latency, otherwise we would be missing some transactions.

shiveshr commented on controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java:
```diff
@@ -222,5 +239,6 @@ public void close() {
         sealStreamLatency.close();
         updateStreamLatency.close();
         truncateStreamLatency.close();
+        INSTANCE.set(null);
```
you are modifying a static variable in the close of an object reference. This may not be a good idea generally speaking. You will end up setting this static variable across close from multiple objects.

shiveshr replied:
Basically dont make this class Closable/AutoClosable. 
Dont let its users call close on it. 
They should call a new `static synchronized reset` method explicitly which takes the existing object and calls a private close on it. 
This will ensure there exists at maximum a single open object and there is never a closed object that getInstance method would return.

RaulGracia replied:
Done.

shiveshr commented on controller/src/main/java/io/pravega/controller/metrics/TransactionMetrics.java:
```diff
@@ -31,9 +38,34 @@
+    }
+
+    public static synchronized TransactionMetrics getInstance() {
+        if (INSTANCE.get() == null) {
```
same as above

shiveshr commented on controller/src/main/java/io/pravega/controller/metrics/TransactionMetrics.java:
```diff
@@ -128,7 +215,13 @@ public static void reportOpenTransactions(String scope, String streamName, int o
         abortTransactionLatency.close();
+        abortTransactionSegmentsLatency.close();
+        abortingTransactionLatency.close();
+        INSTANCE.set(null);
```
same as above

shrids commented on controller/src/main/java/io/pravega/controller/metrics/TransactionMetrics.java:
```diff
@@ -29,11 +36,36 @@
+    }
+
+    public static synchronized TransactionMetrics getInstance() {
+        if (INSTANCE.get() == null) {
```
A `getInstance()` call post `reset()` will create a new instance of TransactionMetrics. This can be avoided.
(Using AutoCloseable interface is also desirable).

shiveshr replied:
the intent of reset is to destroy the previous instantiation. 

Also, auto closable will be on the object. Which would mean the object could be closed and getInstance could return a closed object. Which should not be allowed. 
The close method on the object should be effectively private (logically it is similar to what @RaulGracia has done now).

shrids replied:
ok.
I assumed `reset()` is invoked when we are planning to close the instance of TransactionMetrics i.e. after reset() is invoked we would not want to invoke `getInstance()`. I was thinking more on the lines where calling `getInstance()` post the close would throw an `ObjectClosedException`.

shiveshr commented on controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java:
```diff
@@ -38,13 +39,30 @@
-public final class StreamMetrics extends AbstractControllerMetrics implements AutoCloseable {
+public final class StreamMetrics extends AbstractControllerMetrics {
+
+    private static final AtomicReference<StreamMetrics> INSTANCE = new AtomicReference<>();
```
minor nit: since instance is protected by lock (synchronized methods), we can make it non final and remove atomic reference which is basically double locking.

RaulGracia replied:
I observed that our code checks specifically forbid to create non-final static variables for safety. That's the reason to create an `AtomicReference` there.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/ControllerService.java:
```diff
@@ -328,10 +312,10 @@ private SegmentRange convert(final String scope,
                 }).handle((result, ex) -> {
                     if (ex != null) {
-                        transactionMetrics.createTransactionFailed(scope, stream);
+                        TransactionMetrics.getInstance().createTransactionFailed(scope, stream);
```
do we need to call getInstance each time? Can we call it once and store it in a field in this object and use it for the lifecycle of this object?

shiveshr replied:
Calling getInstance which is synchronized from so many place, and from potentially multiple threads for different workflows, will result in lot of lock contention and waits which could be detrimental for performance. 
In the constructor we can call it once and set it in a local field and use it throughout this class.

RaulGracia replied:
As far as I know, this is how the traditional singleton pattern defines to get the instance. However, if you want, I could "embed" the `getInstance()` call on each method of `TransactionMetrics` and `StreamMetrics`, so from the usage perspective, it would be virtually like calling a static method (i.e., no need to explicitly call `getInstance()`). Is that better? I think that what we want to avoid is to add extra fields to classes just to report these metrics, so created a reference field on each class would be kind of against the change, right?

shiveshr replied:
The concern is everytime you call getInstance you contest for lock. 

Instead, you can call getInstance once when an object of this class was created  -- e.g. ControllerServiceStarter's constructor and keep that reference in a field in ControllerServiceStarter.

So any subsequent call in ControllerServiceStarter for metrics will use the referenced object instead of calling getInstance again.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java:
```diff
@@ -186,6 +184,10 @@ protected void startUp() {
             log.info("Creating the checkpoint store");
             checkpointStore = CheckpointStoreFactory.create(storeClient);
 
+            // Initialize Stream and Transaction metrics.
```
why do we need to do this initialization? 
Upon its first usage, it would get initialized once and reused everywhere.

shiveshr replied:
Although i think its better for the readability of the code.

RaulGracia replied:
Yes, to make the initialization explicit I invoked it there.

kevinhan88 commented:
First of all, this PR is doing two separate things: 
1. add more metrics; 
2. refactor StreamMetrics and TransactionMetrics using Singleton pattern.
I'd suggest next time we address separate concern with separate tickets, in particular the Singleton refactoring is a significant one.

Secondly, I have a few concerns regarding the singleton refactoring.
1. Don't see the need. Even in the worst case there are multiple StreamMetrics or TransactionMetrics objects got created (from existing code I didn't see the case though), it has no harm at all - because the metrics are "step" based, having more objects is like having more steps. It doesn't change functionality.
2. Then what we gain with the overhead of the singleton? Note for every metrics call, it must go through the chain of "getInstance()..."
3. Even more importantly, AutoCloseable is guaranteed by language, while the newly introduced "initialize"/"reset" mechanism is not. There is a risk of memory leaking if the method is not called.

RaulGracia commented:
@kevinhan88 let me address your comments:
1) The fact of also refactoring `StreamMetrics` in this PR is because is a class doing exactly the same and in the same way as `TransactionsMetrics`, so we considered fine to refactor it as well.
2) The Singleton-like refactor of these clases is not because of generating more or less object, is because i) we need a single instance of `StreamMetrics` and `TransactionsMetrics` to prevent opening multiple `OpStatsLogger` with the same name, which are conflicting among them, ii) these clases are used from multiple classes. The way we were doing this was passing an object to each class using `StreamMetrics` and/or `TransactionsMetrics` via a constructor, but this ends up impacting the signature of many constructors, which in turn also impacts testing code and is annoying. 

The approach taken here achieves addresses these points: makes `StreamMetrics` and `TransactionsMetrics` easier to use without impacting the classes where they are used. Also, the lifecycle of the single instances of these clases are managed in a single point (`ControllerServiceStarter`).

RaulGracia commented:
@shiveshr verified that current code works in a cluster and metrics are correctly reported even after controller restarts.
